vm/src/protocol/buffer.rs

```diff
@@ -123,17 +146,293 @@ impl TryFromBorrowedObject for PyBuffer {
 // but it is not supported by Rust
 impl Drop for PyBuffer {
     fn drop(&mut self) {
-        self.internal.release();
+        self.release();
     }
 }
 
-impl Clone for PyBuffer {
-    fn clone(&self) -> Self {
-        self.clone_with_options(self.options.clone())
+#[derive(Debug, Clone)]
+pub struct BufferDescriptor {
+    /// product(shape) * itemsize
+    /// bytes length, but not the length for obj_bytes() even is contiguous
+    pub len: usize,
+    pub readonly: bool,
+    pub itemsize: usize,
+    pub format: Cow<'static, str>,
+    /// (shape, stride, suboffset) for each dimension
+    pub dim_desc: Vec<(usize, isize, isize)>,
+}
+
+impl BufferDescriptor {
+    pub fn simple(bytes_len: usize, readonly: bool) -> Self {
+        Self {
+            len: bytes_len,
+            readonly,
+            itemsize: 1,
+            format: Cow::Borrowed("B"),
+            dim_desc: vec![(bytes_len, 1, 0)],
+        }
     }
+
+    pub fn format(
+        bytes_len: usize,
+        readonly: bool,
+        itemsize: usize,
+        format: Cow<'static, str>,
+    ) -> Self {
+        Self {
+            len: bytes_len,
+            readonly,
+            itemsize,
+            format,
+            dim_desc: vec![(bytes_len / itemsize, itemsize as isize, 0)],
+        }
+    }
+
+    #[cfg(debug_assertions)]
+    pub fn validate(self) -> Self {
+        assert!(self.itemsize != 0);
+        assert!(self.ndim() != 0);
+        let mut shape_product = 1;
+        for (shape, stride, suboffset) in self.dim_desc.iter().cloned() {
+            shape_product *= shape;
+            assert!(suboffset >= 0);
+            assert!(stride != 0);
+            if stride.is_negative() {
+                // if stride is negative, we access memory in reversed order
+                // so the suboffset should be n*stride shift the index to the tail
+                assert!(suboffset >= -stride * shape as isize);
+            }
+        }
+        assert!(shape_product * self.itemsize == self.len);
+        self
+    }
+
+    #[cfg(not(debug_assertions))]
+    pub fn validate(self) -> Self {
+        self
+    }
+
+    pub fn ndim(&self) -> usize {
+        self.dim_desc.len()
+    }
+
+    pub fn is_contiguous(&self) -> bool {
+        if self.len == 0 {
+            return true;
+        }
+        let mut sd = self.itemsize;
+        for (shape, stride, _) in self.dim_desc.iter().cloned().rev() {
+            if shape > 1 && stride != sd as isize {
+                return false;
+            }
+            sd *= shape;
+        }
+        true
+    }
+
+    /// this function do not check the bound
+    /// panic if indices.len() != ndim
+    pub fn get_position_fast(&self, indices: &[usize]) -> isize {
```

Comment:
if the name `get_` is not coming from CPython, I want to suggest to remove it because this function is not perform getter or getitem. Just `fast_position` like `fast_getitem`?